core/platform/email/mailgun_email_services.py

```diff
@@ -14,138 +14,116 @@
-        data['h:Reply-To'] = reply_to
-
-    post_to_mailgun(data)
-
-
-def send_bulk_mail(
-        sender_email, recipient_emails, subject, plaintext_body, html_body):
-    """Sends an email using mailgun api.
-
-    In general this function should only be called from
-    email_manager._send_email().
-
-    Args:
-        sender_email: str. the email address of the sender. This should be in
-            the form 'SENDER_NAME <SENDER_EMAIL_ADDRESS>'.
-        recipient_emails: list(str). list of the email addresses of recipients.
-        subject: str. The subject line of the email.
-        plaintext_body: str. The plaintext body of the email.
-        html_body: str. The HTML body of the email. Must fit in a datastore
-            entity.
-
-    Raises:
-        Exception: if the configuration in feconf.py forbids emails from being
-            sent.
-        Exception: if mailgun api key is not stored in feconf.MAILGUN_API_KEY.
-        Exception: if mailgun domain name is not stored in
-            feconf.MAILGUN_DOMAIN_NAME.
-            (and possibly other exceptions, due to mail.send_mail() failures)
-    """
-    if not feconf.CAN_SEND_EMAILS:
-        raise Exception('This app cannot send emails to users.')
-
     # To send bulk emails we pass list of recipients in 'to' paarameter of
     # post data. Maximum limit of recipients per request is 1000.
     # For more detail check following link:
     # https://documentation.mailgun.com/user_manual.html#batch-sending
-    recipient_email_sets = [
+    recipient_email_lists = [
         recipient_emails[i:i + 1000]
         for i in python_utils.RANGE(0, len(recipient_emails), 1000)]
+    for email_list in recipient_email_lists:
+        data = {
+            'from': sender_email,
+            'subject': subject,
+            'text': plaintext_body,
+            'html': html_body
+        }
+
+        if len(email_list) == 1:
+            data['to'] = email_list[0]
+        else:
+            data['to'] = email_list
+
+        if bcc:
+            if len(bcc) == 1:
+                data['bcc'] = bcc[0]
+            else:
+                data['bcc'] = bcc
+
+        if reply_to:
+            data['h:Reply-To'] = reply_to
 
-    for email_set in recipient_email_sets:
         # 'recipient-variable' in post data forces mailgun to send individual
         # email to each recipient (This is intended to be a workaround for
         # sending individual emails).
-        data = {
-            'from': sender_email,
-            'to': email_set,
-            'subject': subject.encode(encoding='utf-8'),
-            'text': plaintext_body.encode(encoding='utf-8'),
-            'html': html_body.encode(encoding='utf-8'),
-            'recipient-variables': '{}'}
-
-        post_to_mailgun(data)
+        if recipient_variables:
+            data['recipient_variables'] = recipient_variables
+        else:
+            data['recipient_variables'] = {}
```

Comment:
Simplify to `data['recipient_variables'] = recipient_variables or {}` (check that this works)

Or, if you want to be less fancy, have 
```
if recipient_variables is None:
    recipient_variables = {}
```
at the top of the function.